xarray/core/common.py

```diff
@@ -21,12 +22,12 @@ class ImplementsArrayReduce(object):
     def _reduce_method(cls, func, include_skipna, numeric_only):
         if include_skipna:
             def wrapped_func(self, dim=None, axis=None, skipna=None,
-                             keep_attrs=False, **kwargs):
+                             keep_attrs=_set_keep_attrs(False), **kwargs):
```

Comment:
This won't work properly, because default arguments are evaluated when functions are defined instead of when the function is called.

Instead, these should probably be written something like this:
```python
def func(..., keep_attrs=None):
    if keep_attrs is None:
        keep_attrs = _get_keep_attrs(default=False)
```

Reply:
You're right. Updated to be as you suggested everywhere, and tests added to check that the behaviour is as expected.